### PR TITLE
Add SPI support for CC13xx / CC26xx

### PR DESCRIPTION
**Pull request by bwitherspoon**
Add support for the CC13xx / CC26xx SPI peripheral. Support for slave mode and word sizes other than 8 bits has not been implemented.

### CODE REVIEW

**Comment by bwitherspoon** [comment]
The driver allows the CS pin to be configured as hardware (_cs-pin_ attribute in device tree) and / or software (_cs-gpios_ attribute in device tree) controlled which is required for some configurations. If both the _cs-pin_ and _cs-gpios_ attributes are specified, which one is used depends on whether the `spi_config` given by the user contains a GPIO device and pin. I want to mention this because it might appear unusual in the device trees and relies on the behavior of the `spi_context_cs_configure` function so it might not be obvious.

**Review by vanti — CHANGES_REQUESTED** [review]
Looks good to me. Just a couple of minor comments. I am assuming you have a way to test SPI1, given you have coded everything up and it is not exposed on the LaunchPads.

**Comment by bwitherspoon** [comment]
> Looks good to me. Just a couple of minor comments. I am assuming you have a way to test SPI1, given you have coded everything up and it is not exposed on the LaunchPads.

Good question. For testing, I have mostly been using the BOOSTXL-ULPSENSE which has an ADXL362 accelerometer using SPI. I have added it as a shield with a sample on one of my branches: https://github.com/bwitherspoon/zephyr/tree/boostxl_ulpsense.

To test the SPI1 device I just modified the [device tree overlay](https://github.com/bwitherspoon/zephyr/blob/boostxl_ulpsense/boards/shields/boostxl_ulpsense/boostxl_ulpsense.overlay) for the BOOSTXL-ULPSENSE to disable SPI0 and use SPI1 instead. It then looks like:

```
&spi0 {
	status = "disabled";
};

&spi1 {
	status = "ok";
	sck-pin = <10>;
	mosi-pin = <9>;
	miso-pin = <8>;
	cs-pin = <11>;
	cs-gpios = <&gpio0 11 0>;

	adxl362@0 {
		compatible = "adi,adxl362";
		reg = <0>;
		spi-max-frequency = <8000000>;
		int1-gpios = <&gpio0 18 0>;
		label = "ADXL362";
	};

};
```
Then I modified the [project configuration](https://github.com/bwitherspoon/zephyr/blob/boostxl_ulpsense/samples/shields/boostxl_ulpsense/prj.conf) for the BOOSTXL-ULPSENSE sample to also disable SPI0 and enable SPI1 by adding:

```
CONFIG_SPI_0=n
CONFIG_SPI_1=y
```

Then building and running the sample works as expected using the SPI1 device.

```
west build -b cc1352r1_launchxl zephyr/samples/shields/boostxl_ulpsense/
west flash
...
```

File: drivers/spi/spi_cc13xx_cc26xx.c [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	struct spi_context *ctx = &get_dev_data(dev)->ctx;
+	u32_t prot;
+
+	if (spi_context_configured(ctx, config)) {
+		return 0;
+	}
+
+	if (SPI_OP_MODE_GET(config->operation) != SPI_OP_MODE_MASTER) {
+		LOG_ERR("Slave mode is not supported");
+		return -EINVAL;
```
[the remark this answers is not in the document]
Comment (bwitherspoon) [reply drivers/spi/spi_cc13xx_cc26xx.c]: I too am interested in what others have to say. I remember trying to decide between -ENOTSUP and -EINVAL here because I was trying to do the same as other drivers. I have found that the API documentation sometimes specifies what should be returned. For example, the [I2C](https://docs.zephyrproject.org/latest/reference/peripherals/i2c.htmll) API documentation says return -EIO on any error, but I am not sure all drivers really do this. Unfortunately, the [SPI](https://docs.zephyrproject.org/latest/reference/peripherals/spi.html) API documentation just says a negative errno.

File: drivers/spi/spi_cc13xx_cc26xx.c [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+static struct spi_cc13xx_cc26xx_data spi_cc13xx_cc26xx_data_1 = {
+	SPI_CONTEXT_INIT_LOCK(spi_cc13xx_cc26xx_data_1, ctx),
+	SPI_CONTEXT_INIT_SYNC(spi_cc13xx_cc26xx_data_1, ctx),
+};
+
+DEVICE_AND_API_INIT(spi_cc13xx_cc26xx_1, DT_TI_CC13XX_CC26XX_SPI_40008000_LABEL,
+		    spi_cc13xx_cc26xx_init_1, &spi_cc13xx_cc26xx_data_1,
+		    &spi_cc13xx_cc26xx_config_1, POST_KERNEL,
+		    CONFIG_SPI_INIT_PRIORITY, &spi_cc13xx_cc26xx_driver_api);
+#endif /* CONFIG_SPI_0 */
```
[the remark this answers is not in the document]
Comment (bwitherspoon) [reply drivers/spi/spi_cc13xx_cc26xx.c]: Fixed

**Comment by vanti** [comment]
> Then I modified the [project configuration](https://github.com/bwitherspoon/zephyr/blob/boostxl_ulpsense/samples/shields/boostxl_ulpsense/prj.conf) for the BOOSTXL-ULPSENSE sample to also disable SPI0 and enable SPI1 by adding:
> 
> ```
> CONFIG_SPI_0=n
> CONFIG_SPI_1=y
> ```
> 

Would you need to add something to Kconfig to allow CONFIG_SPI_1 to be set? I don't see SPI_1 as a config parameter in the PR.

**Comment by bwitherspoon** [comment]
> Would you need to add something to Kconfig to allow CONFIG_SPI_1 to be set? I don't see SPI_1 as a config parameter in the PR.

The SPI directory level [Kconfig](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/spi/Kconfig) has some generic SPI_* configs I used instead of defining driver specific ones.

**Comment by bwitherspoon** [comment]
I have added PRCMPeripheralSleepEnable() to keep the SPI peripheral enabled while sleeping until proper power management is implemented.

**Comment by vanti** [comment]
> I have added PRCMPeripheralSleepEnable() to keep the SPI peripheral enabled while sleeping until proper power management is implemented.

Thanks. It'd be good to add PRCMPeripheralDeepSleepEnable() as well for deep sleep mode.

**Comment by bwitherspoon** [comment]
I added PRCMPeripheralDeepSleepEnable() and also updated the boards' YAML and documentation to indicate SPI support.

**Review by vanti — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]
doc changes LGTM

**Comment by bwitherspoon** [comment]
Rebased on master.

File: drivers/spi/spi_cc13xx_cc26xx.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/spi/spi_cc13xx_cc26xx.c]: go for -ENOTSUP

File: dts/arm/ti/cc13x2_cc26x2.dtsi [hunk 3]
```diff
@@ -87,6 +87,26 @@
 			status = "disabled";
 			label = "I2C_0";
 		};
+
+		spi0: spi@40000000 {
```
Comment (tbursztyka) [inline dts/arm/ti/cc13x2_cc26x2.dtsi]: Put the dts specific files into a separate commit
